python/thunder/rdds/series.py

```diff
@@ -636,6 +636,111 @@ def query(self, inds, var='inds', order='F', onebased=True):
+        taken from the key of the first Series record in that partition. If the Series object is already
+        sorted and no records have been removed by filtering, then the resulting output should be equivalent
+        to what one would get from calling myImages.saveAsBinarySeries().
+
+        The json config file written out by this method takes the value of the 'nvalues' key from the
+        length of this object's .index attribute. If the index has been modified to have a different length
+        than the number of values in a Series record, then the resulting data files will be read back in
+        incorrectly.
+
+        Parameters
+        ----------
+        outputdirname : string path or URI to directory to be created
+            Output files will be written underneath outputdirname. This directory must not yet exist
+            (unless overwrite is True), and must be no more than one level beneath an existing directory.
+            It will be created as a result of this call.
+
+        overwrite : bool
+            If true, outputdirname and all its contents will be deleted and recreated as part
+            of this call.
+        """
+        import cStringIO as StringIO
+        import struct
+        from thunder.rdds.imgblocks.blocks import SimpleBlocks
+        from thunder.rdds.fileio.writers import getParallelWriterForPath
+        from thunder.rdds.fileio.seriesloader import writeSeriesConfig
+
+        def partitionToBinarySeries(kvIter):
+            """Collects all Series records in a partition into a single binary series record.
+            """
+            keypacker = None
+            firstKey = None
+            buf = StringIO.StringIO()
+            for seriesKey, series in kvIter:
+                if keypacker is None:
+                    keypacker = struct.Struct('h'*len(seriesKey))
+                    firstKey = seriesKey
+                # print >> sys.stderr, seriesKey, series, series.tostring().encode('hex')
+                buf.write(keypacker.pack(*seriesKey))
+                buf.write(series.tostring())
+            val = buf.getvalue()
+            buf.close()
+            # we might have an empty partition, in which case firstKey will still be None
+            if firstKey is None:
+                return iter([])
+            else:
+                label = SimpleBlocks.getBinarySeriesNameForKey(firstKey) + ".bin"
+                return iter([(label, val)])
+
+        writer = getParallelWriterForPath(outputdirname)(outputdirname, overwrite=overwrite)
+
+        binseriesrdd = self.rdd.mapPartitions(partitionToBinarySeries)
+
+        binseriesrdd.foreach(writer.writerFcn)
+        # note that we use len(self.index) here for nvalues! we don't really have a great alternative
+        # to this at the moment, short of calling first(). 
+        writeSeriesConfig(outputdirname, len(self.dims), len(self.index), dims=self.dims.count,
```

Comment:
Ok, two thoughts, one fairly concrete and the other more theoretical:

* `dims` is currently referenced in the config in two places, once as `dims.count` to get the dimension tuple itself, and the other as `len(dims)` to get the number of keys. Number of keys we do have to have here, because otherwise the data is unreadable, but it isn't stored anywhere else in the Series metadata. So our alternative, if `dims` isn't set on the series object, would be to get `nkeys` from a `series.first()` call. Now this is still better than recalculating the full `dims`, since `dims` requires a full map/reduce cycle through the data, but in all likelihood it still won't be instantaneous. (Another plus is that by calling `first`, we'd get to set `nrecords` directly, rather than by assuming that the `.index` attribute is correct.) So I think I'm in favor of this overall, but read on...

* The Series binary format and logic is growing increasingly hairy, for what was originally I think intended and expected to be a really simple nearly-flat-file representation of the data. We're now proposing to make one of the metadata fields (`dims`) optional. This is not in and of itself a big deal, but it does suggest, at least to me, that at some point we may wish to stop and revisit the entire issue of a persisted representation of Series objects, and see if we can't come up with some more principled, generic, and reusable format, rather than continuing to make ad-hoc changes. (For some reason I keep wanting to say "protocol buffers! avro! thrift!" here, but I'm by no means convinced that's a good idea either. :) )

Takeaway: sure, I'll make `dims` optional, just so long as we're aware that we still have to make a `first()` call in this case.